metrics/framework/Metrics.cpp

```diff
@@ -0,0 +1,212 @@
+{
+    mv.emplace_back((std::make_shared<Measurement<uint32_t>>(name, description, ValueType::METRICS_INTEGER, count.load())));
+}
+
+
+void GaugeMetric::getReportValues(MeasurementVector &mv) const
+{
+    mv.emplace_back((std::make_shared<Measurement<float>>(name, description, ValueType::METRICS_FLOAT, gaugeValue.load())));
+}
+
+
+IMetricSink *MetricSink::getSinkFromLib(const char *type, const char *sinkName, const IPropertyTree *pSettingsTree)
+{
+    return MetricSink::getSinkFromLib(type, "", sinkName, pSettingsTree);
+}
+
+IMetricSink *MetricSink::getSinkFromLib(const char *type, const char *getInstanceProcName, const char *sinkName, const IPropertyTree *pSettingsTree)
+{
+    std::string libName;
+
+    //
+    // First, treat type as a full library name
+    libName = type;
+    if (libName.find(SharedObjectExtension) == std::string::npos)
+    {
+        libName.append(SharedObjectExtension);
+    }
+    HINSTANCE libHandle = LoadSharedObject(libName.c_str(), true, false);
+
+    //
+    // If type wasn't the lib name, treat it as a type
+    if (libHandle == nullptr)
+    {
+        libName.clear();
+        libName.append("libhpccmetrics_").append(type).append(SharedObjectExtension);
+        libHandle = LoadSharedObject(libName.c_str(), true, false);
+    }
+
+    //
+    // If able to load the lib, get the instance proc and create the sink instance
+    IMetricSink *pSink = nullptr;
+    if (libHandle != nullptr)
+    {
+        const char *epName = !isEmptyString(getInstanceProcName) ? getInstanceProcName : "getSinkInstance";
+        auto getInstanceProc = (getSinkInstance) GetSharedProcedure(libHandle, epName);
+        if (getInstanceProc != nullptr)
+        {
+            const char *name = isEmptyString(sinkName) ? type : sinkName;
+            pSink = getInstanceProc(name, pSettingsTree);
+        }
+    }
+    return pSink;
+}
+
+
+void MetricsReporter::init(IPropertyTree *pMetricsTree)
+{
+    initializeSinks(pMetricsTree->getElements("sinks"));
+}
+
+
+void MetricsReporter::addMetric(const std::shared_ptr<IMetric> &pMetric)
+{
+    std::unique_lock<std::mutex> lock(metricVectorMutex);
+    auto it = metrics.find(pMetric->getName());
+    if (it == metrics.end())
+    {
+        metrics.insert({pMetric->getName(), pMetric});
+    }
+    // note that
+}
+
+
+void MetricsReporter::removeMetric(const std::string &metricName)
+{
+    std::unique_lock<std::mutex> lock(metricVectorMutex);
+    auto it = metrics.find(metricName);
+    if (it != metrics.end())
+    {
+        metrics.erase(it);
+    }
+}
+
+
+void MetricsReporter::startCollecting()
+{
+    for (auto const &sinkIt : sinks)
+    {
+        sinkIt.second.pSink->startCollection(this);
+    }
+}
+
+
+void MetricsReporter::stopCollecting()
+{
+    for (auto const &sinkIt : sinks)
+    {
+        sinkIt.second.pSink->stopCollection();
+    }
+}
+
+
+std::vector<std::shared_ptr<IMetric>> MetricsReporter::getReportMetrics(const std::string &sinkName)
```

Comment:
This seems to add an unnecessary extra step - create a list of metrics to report, and then call report on them.  There should be a function to gather the measurements, without returning the temporary list.